### PR TITLE
Add file backend support for Blob and related

### LINKED ISSUE #10851: Add file handling functionality to Blob object
**NOTE**: This is an issue assigned to myself

Spec: https://w3c.github.io/FileAPI/#blob

As noted before, current Blob object can only refer to in-memory data ([`DataSlice`](https://doc.servo.org/script/dom/blob/struct.DataSlice.html)), and the [`FileReader`](https://doc.servo.org/script/dom/filereader/struct.FileReader.html) will also only operate in this way. In order to support reading from file system, another internal data structure might need to be added to the `Blob`. It might be a file handler created at the time of object construction.

About snapshot state:

> If a File object is a reference to a byte sequence originating from a file on disk, then its snapshot state should be set to the state of the file on disk at the time the File object is created.
> 
> Note: This is a non-trivial requirement to implement for user agents, and is thus not a must but a should [RFC2119]. User agents should endeavor to have a File object’s snapshot state set to the state of the underlying storage on disk at the time the reference is taken. If the file is modified on disk following the time a reference has been taken, the File's snapshot state will differ from the state of the underlying storage. User agents may use modification time stamps and other mechanisms to maintain snapshot state, but this is left as an implementation detail.

Working at [branching here](https://github.com/izgzhen/servo/tree/blob)

### PR DESCRIPTION
**Pull request by izgzhen**
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy --faster` does not report any errors
- [x] These changes fix #10851, related to #11131
- [x] These changes do not require tests because the implementation is partial and can't work alone
### Major changes
1. Add new backend to `Blob` and a `BlobImpl` struct to abstract multiple backends
2. Rewrite most interfaces of `Blob` to accommodate the change
3. Change the `read` behaviour of `FileReader`, considering the case when blob is file-backed and not cached

The design is still immature, welcome comments!
### Problems to resolve
- [x] I used `DOMRefCell` to cache the bytes in `BlobImpl`, is it sound?
- [x] The interfaces (like `BlobImpl::get_bytes`) handle requests in a default-to-empty way when the inner `DataSlice` is not cached. It might be possible to handle this condition better.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11221)

### CODE REVIEW

**Comment by izgzhen** [comment]
@Manishearth I refined the design a bit and I think we can push the progress of this thread in parallel with #11189

File: components/script/dom/blob.rs [hunk 1]
```diff
@@ -88,22 +90,76 @@ impl DataSlice {
...
     #[ignore_heap_size_of = "No clear owner"]
-    data: DataSlice,
+    blob_impl: BlobImpl,
     typeString: String,
     isClosed_: Cell<bool>,
 }
 
+#[derive(Clone, JSTraceable)]
+pub struct BlobImpl {
+    slice: DOMRefCell<Option<DataSlice>>,
```
Comment (Manishearth) [inline components/script/dom/blob.rs]: Perhaps this should be an enum?

File: components/script/dom/blob.rs [hunk 2]
```diff
@@ -88,22 +90,76 @@ impl DataSlice {
...
+            file_id: Some(file_id),
+        }
+    }
+
+    pub fn new_from_empty_slice() -> BlobImpl {
+        BlobImpl::new_from_slice(DataSlice::empty())
+    }
+
+    pub fn get_bytes(&self) -> Vec<u8> {
+        match *self.slice.borrow() {
```
Comment (Manishearth) [inline components/script/dom/blob.rs]: what if it's a file?

File: components/script/dom/blob.rs (hunk 1 above)
Comment (izgzhen) [inline components/script/dom/blob.rs]: I am thinking about the possibility of caching here. If it is file-backed, we can put the cached content inside the other `DataSlice` field.

Comment (Manishearth) [inline components/script/dom/blob.rs]: I'd rather have an enum here and within the FileId variant have an option for caching. As it stands slice = Some can mean cached or an in-memory slice, a clear boundary would be nicer and more Rusty.

File: components/script/dom/blob.rs (hunk 2 above)
Comment (izgzhen) [inline components/script/dom/blob.rs]: Currently I don't put any read operations in `blob.rs`, which means that the user needs to check if it is cached and then cache if necessary. Or, if the file is not cached, a empty slice or stuff with similar semantics will be returned.

Comment (Manishearth) [inline components/script/dom/blob.rs]: I feel like blob.rs should handle get_bytes either way. Of course it shouldn't be the one doing the reading, but it should talk to the resource task and fetch the reads and handle caching.

Comment (izgzhen) [inline components/script/dom/blob.rs]: It am not sure if it is possible to fetch a reference to the `window` within `Blob` itself, which means that I can't access script_thread stuff.

Comment (Manishearth) [inline components/script/dom/blob.rs]: `blob.global()` should work

**Comment by izgzhen** [comment]
@Manishearth In order to get port to filemanger thread inside blob (by `GlobalRef`), I need to attach the `filemanager_thread` to worker (in `workerglobalscope.rs`). Currently the worker only has access to `core_resource_thread` (renamed from original `resource_thread`), rather than the combined `ResourceThreads` from which I can access the file manager. However I don't know if storage API (`storage_thread` I think) can be exposed to worker as well.

**Comment by jdm** [comment]
Exposing it won't hurt anything; the storage APIs aren't accessible from workers, so there hasn't been any need to make the storage thread available.

**Comment by Manishearth** [comment]
Any particular reason for workers not being able to use storage? If there's a security reason we should not even give workers an unused storage chan (sandboxing).

**Comment by Manishearth** [comment]
Seems like it [used to be for security reasons](https://bugs.chromium.org/p/chromium/issues/detail?id=105495) but now it's just for "threading issues".

**Comment by izgzhen** [comment]
I didn't see any `Exposed` or similar in this spec http://www.w3.org/TR/2016/REC-webstorage-20160419/#the-storage-interface

**Comment by Manishearth** [comment]
Oh, sure, the spec doesn't let you, the question is: If a sandboxed worker was compromised with an RCE vulnerability and got access to storage data, is that an escalation of access?

Assuming that we have checks in place in the resource/storage thread that ensure that no process gets data from a different origin (we don't, but we should eventually), I don't think this is an issue. Also, I don't see us putting workers in separate processes, so if a worker thread is compromised the script thread would be compromised too anyway.

File: components/script/dom/blob.rs [hunk 3]
```diff
@@ -121,23 +155,53 @@ impl Blob {
...
 }
 
+fn read_file(global: GlobalRef, id: Uuid) -> DataSlice {
+    // FIXME: After resource thread #11189 PR is landed
+    // let global = global.filemanager_thread();
+
+    let file_manager: IpcSender<FileManagerThreadMsg> = unimplemented!();
+    let (chan, recv) = ipc::channel().unwrap();
+    let _ = file_manager.send(FileManagerThreadMsg::ReadFile(chan, id));
+    let result = recv.recv().unwrap();
```
Comment (Manishearth) [inline components/script/dom/blob.rs]: Should we be handling/propagating errors here?

File: components/script/dom/htmlformelement.rs [hunk 4]
```diff
@@ -313,7 +313,8 @@ impl HTMLFormElement {
                         content_disposition,
                         content_type));
 
-                    result.push_str(from_utf8(&f.upcast::<Blob>().get_data().get_bytes()).unwrap());
+                    let slice = f.upcast::<Blob>().get_slice();
+                    result.push_str(from_utf8(&slice.get_bytes()).unwrap());
```
Comment (Manishearth) [inline components/script/dom/htmlformelement.rs]: Not necessarily utf8 here? The charset can change.

**Comment by Manishearth** [comment]
Almost done!

File: components/script/dom/blob.rs (hunk 3 above)
Comment (izgzhen) [inline components/script/dom/blob.rs]: I think `read_file` can be made `Option<DataSlice>` and we can throw a script-level exception or something at the dom API where it is used (like `Slice`)

Comment (Manishearth) [inline components/script/dom/blob.rs]: `Result<DataSlice,()>` and use `try!` or the `?` operator, but yeah

File: components/script/dom/htmlformelement.rs [hunk 5]
```diff
@@ -308,12 +308,17 @@ impl HTMLFormElement {
                         DispositionParam::Filename(Charset::Ext(String::from(charset.clone())),
                                                    None,
                                                    f.name().clone().into()));
+                    /// XXX: unwrap
```
Comment (Manishearth) [inline components/script/dom/htmlformelement.rs]: replace it with an expect

though in this case `unwrap_or(text/plain)` might be better?

Comment (izgzhen) [inline components/script/dom/htmlformelement.rs]: Not sure, can't find related spec, maybe an empty `""`?

Comment (Manishearth) [inline components/script/dom/htmlformelement.rs]: Section 4.4, it defaults to text/plain

https://tools.ietf.org/html/rfc7578#section-4.4

File: components/script/dom/blob.rs [hunk 6]
```diff
@@ -118,35 +152,70 @@ impl Blob {
...
-                &BlobOrString::Blob(ref b) => {
-                    b.get_data().get_bytes().to_vec()
-                },
-            }
-        }).collect::<Vec<u8>>()
+fn read_file(global: GlobalRef, id: Uuid) -> Result<DataSlice, ()> {
+    let file_manager = global.filemanager_thread();
+    let (chan, recv) = ipc::channel().map_err(|_|())?;
+    let _ = file_manager.send(FileManagerThreadMsg::ReadFile(chan, id));
+    let result = recv.recv().unwrap();
```
Comment (Manishearth) [inline components/script/dom/blob.rs]: This should propagate an error

File: components/script/dom/blob.rs [hunk 7]
```diff
@@ -118,35 +152,70 @@ impl Blob {
...
+    /// XXX: it is still possible to bring down the whole program, but the API might be
+    ///      intuitive to use (since we read FS and cache in a transparent style)
+    pub fn get_slice(&self) -> DataSlice {
+        match self.blob_impl {
+            BlobImpl::File(ref id, ref slice) => {
+                match *slice.borrow() {
+                    Some(ref s) => s.clone(),
+                    None => {
+                        let global = self.global();
+                        let s = read_file(global.r(), id.clone()).expect("Read file in get_slice failed");
```
Comment (Manishearth) [inline components/script/dom/blob.rs]: This should return a Result, too, and callers should pretend that there are no bytes in the Err case.

Comment (izgzhen) [inline components/script/dom/blob.rs]: So users use `unwrap_or`? Looks not bad, I will try it

File: components/script/dom/blob.rs [hunk 8]
```diff
@@ -118,35 +152,81 @@ impl Blob {
...
-        }).collect::<Vec<u8>>()
+fn read_file(global: GlobalRef, id: Uuid) -> Result<DataSlice, ()> {
+    let file_manager = global.filemanager_thread();
+    let (chan, recv) = ipc::channel().map_err(|_|())?;
+    let _ = file_manager.send(FileManagerThreadMsg::ReadFile(chan, id));
+
+    let result = match recv.recv() {
+        Ok(ret) => ret,
+        Err(e) => {
+            debug!("File manager thread has problem {:?}", e);
```
Comment (izgzhen) [inline components/script/dom/blob.rs]: @Manishearth it won't crash now I think, and more explicit.

**Comment by Manishearth** [comment]
@bors-servo r+

thanks!

**Comment by Manishearth** [comment]
@bors retry
 - IPC Arc issue

**Comment by izgzhen** [comment]
Thanks, that's quite spectacular :$

**Comment by Manishearth** [comment]
@bors-servo r-

blocked on https://github.com/servo/ipc-channel/pull/77

**Comment by Manishearth** [comment]
Could you update ipc-channel to 0.2.3 while you're at it?

**Comment by izgzhen** [comment]
@Manishearth Great job! I will do it

**Comment by Manishearth** [comment]
@bors retry try-
 - interrupt

**Pull request merged**